File: packages/arcantry/src/changelog.ts
import { rcompare, valid } from 'semver';

export const changelogCategories = ['Added', 'Changed', 'Deprecated', 'Removed', 'Fixed', 'Security'] as const;
export type ChangelogCategory = (typeof changelogCategories)[number];

export type ChangelogRelease = {
  version: string;
  date: string;
  summary?: string;
  categories: Partial<Record<ChangelogCategory, string[]>>;
};

export type ChangelogInspection = {
  format: 'keep-a-changelog-2' | 'keep-a-changelog-1' | 'compatible' | 'unknown';
  releases: string[];
  hasUnreleased: boolean;
  diagnostics: string[];
};

const versionHeading = /^## \[(\d+\.\d+\.\d+)\] - (\d{4}-\d{2}-\d{2})\s*$/gm;
const anyReleaseHeading = /^## \[\d+\.\d+\.\d+\] - \d{4}-\d{2}-\d{2}\s*$/m;
const unreleasedHeading = /^## \[Unreleased\]\s*$/m;

export const keepAChangelogPreamble = [
  '# Changelog',
  '',
  'All notable changes to this project will be documented in this file.',
  '',
  'The format is based on [Keep a Changelog](https://keepachangelog.com/en/2.0.0/),',
  'and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).',
  '',
].join('\n');

export const inspectChangelog = (content: string): ChangelogInspection => {
  const plain = content.replace(/^\uFEFF/, '');
  const releases = [...plain.matchAll(versionHeading)].map((match) => match[1]!);
  const diagnostics: string[] = [];
  if (!plain.startsWith('# Changelog')) diagnostics.push('Changelog must start with # Changelog.');
  if (!unreleasedHeading.test(plain)) diagnostics.push('Changelog has no [Unreleased] section.');

  for (const heading of plain.matchAll(/^### (.+)$/gm)) {
    if (!changelogCategories.includes(heading[1] as ChangelogCategory)) {
      diagnostics.push(`Unknown Keep a Changelog category: ${heading[1]}.`);
    }
  }

  const format = plain.includes('keepachangelog.com/en/2.0.0')
    ? 'keep-a-changelog-2'
    : /keepachangelog\.com\/en\/1\./.test(plain)
      ? 'keep-a-changelog-1'
      : releases.length > 0 || unreleasedHeading.test(plain)
        ? 'compatible'
        : 'unknown';
  return { format, releases, hasUnreleased: unreleasedHeading.test(plain), diagnostics };
};

export const renderKeepAChangelogV2 = (input: {
  releases: ChangelogRelease[];
  unreleased?: Partial<Record<ChangelogCategory, string[]>>;
  repositoryUrl?: string;
  tagPrefix?: string;
}): string => {
  const tagPrefix = input.tagPrefix ?? 'v';
  const releases = [...input.releases].sort((left, right) => rcompare(left.version, right.version));
  for (const release of releases) {
    if (valid(release.version) === null) throw new Error(`Invalid changelog version: ${release.version}`);
    if (!/^\d{4}-\d{2}-\d{2}$/.test(release.date)) throw new Error(`Invalid changelog date: ${release.date}`);
  }

  const sections = [keepAChangelogPreamble.trimEnd(), '## [Unreleased]', renderCategories(input.unreleased ?? {})];
  for (const release of releases) {
    sections.push(
      `## [${release.version}] - ${release.date}`,
      release.summary?.trim() ?? '',
      renderCategories(release.categories),
    );
  }

  const body = `${sections.filter((section) => section.length > 0).join('\n\n')}\n`;
  if (input.repositoryUrl === undefined || releases.length === 0) return body;
  const repositoryUrl = input.repositoryUrl.replace(/\/$/, '');
  const links = [`[Unreleased]: ${repositoryUrl}/compare/${tagPrefix}${releases[0]!.version}...HEAD`];
  for (const [index, release] of releases.entries()) {
    const previous = releases[index + 1];
    links.push(
      previous === undefined
        ? `[${release.version}]: ${repositoryUrl}/releases/tag/${tagPrefix}${release.version}`
        : `[${release.version}]: ${repositoryUrl}/compare/${tagPrefix}${previous.version}...${tagPrefix}${release.version}`,
    );
  }
  return `${body}\n${links.join('\n')}\n`;
};

export const cutoverChangelog = (content: string, managedFrom: string): string => {
  if (valid(managedFrom) === null) throw new Error('managed_from must be a full SemVer version for changelog cutover.');
  if (content.includes(`Arcantry manages releases from ${managedFrom}.`)) return content;
  const plain = content.replace(/^\uFEFF/, '');
  const firstRelease = plain.search(anyReleaseHeading);
  if (firstRelease < 0) throw new Error('Changelog cutover requires at least one dated version section.');

  const unreleased = plain.match(unreleasedHeading);
  if (unreleased !== null && unreleased.index !== undefined && unreleased.index < firstRelease) {
    const bodyStart = unreleased.index + unreleased[0].length;
    const unreleasedBody = plain.slice(bodyStart, firstRelease).replace(/^###\s+.*$/gm, '').trim();
    if (unreleasedBody.length > 0) {
      throw new Error('Changelog cutover cannot preserve a non-empty Unreleased section without explicit OpenSpec meaning.');
    }
  }

  const bom = content.startsWith('\uFEFF') ? '\uFEFF' : '';
  const legacy = plain.slice(firstRelease);
  const marker = `<!-- Arcantry manages releases from ${managedFrom}. Earlier history remains project-owned. -->`;
  return `${bom}${keepAChangelogPreamble}\n## [Unreleased]\n\n${marker}\n\n${legacy}`;
};

export const migrateChangelogToV2 = (content: string): string => {
  const inspection = inspectChangelog(content);
  if (inspection.format === 'keep-a-changelog-2') return content;
  if (inspection.format !== 'keep-a-changelog-1') {
    throw new Error('Full changelog migration requires an identified Keep a Changelog 1.x source.');
  }
  return content.replace(/https:\/\/keepachangelog\.com\/en\/1(?:\.\d+){1,2}\/?/g, 'https://keepachangelog.com/en/2.0.0/');
};

const renderCategories = (categories: Partial<Record<ChangelogCategory, string[]>>): string =>
  changelogCategories
    .flatMap((category) => {
      const entries = categories[category]?.map((entry) => entry.trim()).filter(Boolean) ?? [];
      return entries.length === 0 ? [] : [`### ${category}`, '', ...entries.map((entry) => `- ${entry}`), ''];
    })
    .join('\n')
    .trimEnd();

File: crates/arcantry-core/src/changelog.rs
use anyhow::{Result, bail};
use semver::Version;

pub const KEEP_A_CHANGELOG_PREAMBLE: &str = "# Changelog\n\nAll notable changes to this project will be documented in this file.\n\nThe format is based on [Keep a Changelog](https://keepachangelog.com/en/2.0.0/),\nand this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).\n";

pub fn render_empty() -> String {
  format!("{KEEP_A_CHANGELOG_PREAMBLE}\n## [Unreleased]\n")
}

pub fn cutover(content: &str, managed_from: &str) -> Result<String> {
  Version::parse(managed_from).map_err(|_| {
    anyhow::anyhow!("managed_from must be a full SemVer version for changelog cutover.")
  })?;
  if content.contains(&format!("Arcantry manages releases from {managed_from}.")) {
    return Ok(content.to_owned());
  }

  let plain = content.strip_prefix('\u{feff}').unwrap_or(content);
  let first_release = release_heading_offsets(plain)
    .into_iter()
    .next()
    .ok_or_else(|| {
      anyhow::anyhow!("Changelog cutover requires at least one dated version section.")
    })?;
  if let Some(unreleased) = plain.find("## [Unreleased]")
    && unreleased < first_release
  {
    let body = &plain[unreleased + "## [Unreleased]".len()..first_release];
    let remaining = body
      .lines()
      .filter(|line| !line.starts_with("### "))
      .collect::<Vec<_>>()
      .join("\n");
    if !remaining.trim().is_empty() {
      bail!(
        "Changelog cutover cannot preserve a non-empty Unreleased section without explicit OpenSpec meaning."
      );
    }
  }

  let bom = if content.starts_with('\u{feff}') {
    "\u{feff}"
  } else {
    ""
  };
  let marker = format!(
    "<!-- Arcantry manages releases from {managed_from}. Earlier history remains project-owned. -->"
  );
  Ok(format!(
    "{bom}{KEEP_A_CHANGELOG_PREAMBLE}\n## [Unreleased]\n\n{marker}\n\n{}",
    &plain[first_release..]
  ))
}

pub fn migrate_to_v2(content: &str) -> Result<String> {
  if content.contains("keepachangelog.com/en/2.0.0") {
    return Ok(content.to_owned());
  }
  if !content.contains("keepachangelog.com/en/1.") {
    bail!("Full changelog migration requires an identified Keep a Changelog 1.x source.");
  }

  let mut desired = content.to_owned();
  while let Some(start) = desired.find("https://keepachangelog.com/en/1.") {
    let suffix = &desired[start..];
    let end = suffix
      .find(|character: char| {
        character.is_whitespace() || matches!(character, ')' | ']' | '>' | '"' | '\'')
      })
      .unwrap_or(suffix.len());
    desired.replace_range(start..start + end, "https://keepachangelog.com/en/2.0.0/");
  }
  Ok(desired)
}

fn release_heading_offsets(content: &str) -> Vec<usize> {
  content
    .match_indices("## [")
    .filter_map(|(offset, _)| {
      let line = content[offset..].lines().next()?;
      let body = line.strip_prefix("## [")?;
      let (version, date) = body.split_once("] - ")?;
      if Version::parse(version).is_ok()
        && date.len() == 10
        && date.chars().enumerate().all(|(index, value)| {
          if matches!(index, 4 | 7) {
            value == '-'
          } else {
            value.is_ascii_digit()
          }
        })
      {
        Some(offset)
      } else {
        None
      }
    })
    .collect()
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn migrates_keep_a_changelog_one_links() {
    let content = "# Changelog\nhttps://keepachangelog.com/en/1.1.0/\n";
    assert_eq!(
      migrate_to_v2(content).unwrap(),
      "# Changelog\nhttps://keepachangelog.com/en/2.0.0/\n"
    );
  }

  #[test]
  fn cuts_over_legacy_history() {
    let content = "# Changelog\n\n## [Unreleased]\n\n## [1.0.0] - 2026-01-01\n\n- Legacy\n";
    let desired = cutover(content, "1.0.0").unwrap();
    assert!(desired.contains("Arcantry manages releases from 1.0.0."));
    assert!(desired.ends_with("## [1.0.0] - 2026-01-01\n\n- Legacy\n"));
  }
}
